### PR TITLE
fix(drive): invalid protocol version is using to deserialize state

### PR DESCRIPTION
**Pull request by shumkov**
## Issue being fixed or feature implemented

Currently we use latest protocol version to deserialise state. It works until current protocol version == latest protocol version, which obviously won't be always the case.

## What was done?

- Use current protocol version to deserialise the state

## How Has This Been Tested?

None

## Breaking Changes

None

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [x] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/rs-drive-abci/src/execution/storage/fetch_execution_state/mod.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+use crate::error::Error;
+use crate::platform_types::platform_state::PlatformState;
+use dpp::version::PlatformVersion;
+use drive::drive::Drive;
+use drive::query::TransactionArg;
+
+mod v0;
+
+/// Fetches execution state from grovedb storage
+pub fn fetch_execution_state(
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/storage/fetch_execution_state/mod.rs:11]: This should be in Drive instead based on it's parameters

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

File: packages/rs-drive-abci/src/execution/storage/fetch_execution_state/mod.rs (hunk 1 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/storage/fetch_execution_state/mod.rs:11]: Drive doesn't know about execution state

Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/storage/fetch_execution_state/mod.rs:11]: Yes, what I mean... is that Drive should do it and return bytes. And then ABCI can have a method to deserialize the bytes.

File: packages/rs-drive-abci/src/execution/engine/finalize_block_proposal/v0/mod.rs [hunk 2]
```diff
@@ -217,7 +217,7 @@ where
         }
         .into();
 
-        self.update_state_cache(extended_block_info, transaction, platform_version)?;
+        self.update_execution_state(extended_block_info, transaction, platform_version)?;
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/engine/finalize_block_proposal/v0/mod.rs:220]: It is a state cache, it exists on disk, and can be recreated. I do not know what execution state means. Please revert.

[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/engine/finalize_block_proposal/v0/mod.rs]: Done

File: packages/rs-drive-abci/src/execution/storage/fetch_execution_state/v0/mod.rs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+use drive::query::TransactionArg;
+
+pub(super) fn fetch_execution_state_v0(
+    drive: &Drive,
+    transaction: TransactionArg,
+    platform_version: &PlatformVersion,
+) -> Result<Option<PlatformState>, Error> {
+    let maybe_bytes = drive
+        .fetch_execution_state_bytes(transaction, platform_version)
+        .map_err(Error::Drive)?;
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/storage/fetch_execution_state/v0/mod.rs:15]: you should just .map here, instead of the next few lines.

File: packages/rs-drive-abci/src/execution/platform_events/block_end/update_state_cache/v0/mod.rs [hunk 4]
```diff
@@ -46,25 +46,25 @@ where
...
         //todo: verify this with an update
-        PlatformVersion::set_current(PlatformVersion::get(platform_version.protocol_version)?);
+        let version = PlatformVersion::get(platform_version.protocol_version)?;
 
-        // Persist ephemeral data
-        self.store_ephemeral_state(&state_cache, transaction, platform_version)?;
+        PlatformVersion::set_current(version);
+
+        // Persist execution state
+        self.store_execution_state(&state, Some(transaction), platform_version)?;
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/platform_events/block_end/update_state_cache/v0/mod.rs:67]: It doesn't make sense to me since ephemeral means not persisted.

File: packages/rs-drive-abci/src/execution/storage/fetch_execution_state/v0/mod.rs (hunk 3 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/storage/fetch_execution_state/v0/mod.rs]: Good point! Thank you

File: packages/rs-drive-abci/src/execution/platform_events/block_end/update_state_cache/v0/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/platform_events/block_end/update_state_cache/v0/mod.rs:67]: Chat GPT thinks the same:
```
I have a structure in my Rust blockchain project what I called ephemeral. What is means for you? What properties such stucture should have? I want to store this ephemeral structure on disk? Is it still be ephemeral?
```

The term "ephemeral" in the context of computer science and blockchain typically refers to something that is transient or temporary. This means that it is not meant to have a long lifespan and is often used for temporary storage, computations, or data transfer. In the context of your Rust blockchain project, an ephemeral structure would likely have the following properties:

1. **Temporary Lifecycle**: The structure's existence is limited to a specific duration or until a certain condition is met. Once its purpose is fulfilled, it should be discarded or reset.

2. **Limited Scope of Use**: It's typically used for a specific, narrow purpose such as temporary calculations, holding intermediate data, or for communication between components.

3. **In-Memory Storage**: Generally, ephemeral data is stored in memory (RAM) and not meant to be persisted to long-term storage devices like disk drives. This is because in-memory storage is faster and the temporary nature of ephemeral data doesn't necessitate long-term storage.

4. **Volatile**: The data in the structure is expected to be volatile. This means that it can be lost or reset when the program exits or when the data is no longer needed.

5. **Security and Privacy Considerations**: Since ephemeral data can include sensitive information (like cryptographic keys in a blockchain context), it often has built-in mechanisms for secure handling and disposal.

Regarding storing an ephemeral structure on disk, this somewhat contradicts the traditional notion of 'ephemeral'. When you store something on disk, it implies a level of persistence and durability that is not typically associated with ephemeral data. However, there can be scenarios where you might want to temporarily store data on disk, for example:

- **Swap Space Usage**: When there's insufficient memory, some ephemeral data might be swapped to disk temporarily.
- **Performance Optimization**: In some cases, writing to disk temporarily might be faster or more efficient than keeping everything in memory, especially for large datasets.

However, once you store something on disk, you need to ensure that it is securely managed and properly deleted or encrypted as needed, especially in blockchain applications where security is paramount. Additionally, if the data is to be truly ephemeral, there should be mechanisms to ensure its deletion or invalidation after its intended use.

[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/platform_events/block_end/update_state_cache/v0/mod.rs:67]: I am in blockchain, I have a platform state that represents the metadata state of the blockchain. Every block this state changes, during block execution I have named my two states: state and ephemeral state. At the end of the block if the block is accepted the ephemeral state is crushed on top of state. Is ephemeral state an okay name for this?

ChatGPT
In the context of blockchain technology, naming conventions are quite important as they convey the function and nature of the elements within the system. The terms you choose should accurately describe the role and behavior of the components in the blockchain.

In your case, the term "ephemeral state" seems appropriate for a few reasons:

Ephemerality: The word "ephemeral" suggests something that is temporary or transient. This aligns well with the role of this state, as it exists only for the duration of a block's execution and is not permanently recorded on the blockchain until the block is accepted.

Clarity and Intuitiveness: The term "ephemeral state" is intuitively understood in the context of software and data management. It clearly indicates a state that is temporary and subject to change or dismissal.

Distinction from Permanent State: By naming the other state simply as "state", you've created a clear distinction between the permanent, persistent state of the blockchain and the temporary state used during block processing.

However, consider a few points:

Consistency with Existing Terminology: Ensure that the term "ephemeral state" aligns with any existing terminology used in your blockchain platform or the wider blockchain community. Consistency in terminology helps in understanding and collaboration.
Understanding by Target Audience: Make sure that your target audience, whether they are developers, users, or stakeholders, can easily understand and differentiate between "state" and "ephemeral state". If the term might cause confusion, consider alternatives or provide clear documentation.
In summary, "ephemeral state" is a fitting and descriptive term for a state that exists temporarily during the processing of a block and is subject to change until the block is accepted, at which point it merges with the main, persistent state. Just ensure that this naming convention aligns well with the overall terminology and understanding of your platform's users and contributors.

[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/platform_events/block_end/update_state_cache/v0/mod.rs:67]: I think you are giving not enough information and wrong input. The `store_ephemeral_state ` function accept both platform state instances, the one called platform state and another block state that you are referencing as a ephemeral.

[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/platform_events/block_end/update_state_cache/v0/mod.rs:67]: Agreed to call it `store_platfrom_state` and `fetch_platfrom_state`

File: packages/rs-drive-abci/src/execution/engine/initialization/init_chain/v0/mod.rs [hunk 5]
```diff
@@ -79,6 +79,24 @@ where
...
+            &platform_version.drive,
+        )?;
+
+        self.drive.apply_batch_low_level_drive_operations(
+            None,
+            Some(transaction),
+            batch_operations,
+            &mut vec![],
+            &platform_version.drive,
+        )?;
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/engine/initialization/init_chain/v0/mod.rs:98]: Operations are for a batch of things you want to do, for every operations call there should also be a direct call. So we should add a store_current_protocol_version(, also we should name it store instead of set, as it's not a setter and can be confusing for some.

File: packages/rs-drive-abci/src/execution/storage/fetch_platform_state/mod.rs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        .fetch_platform_state
+    {
+        0 => v0::fetch_platform_state_v0(drive, transaction, platform_version),
+        version => Err(Error::Execution(ExecutionError::UnknownVersionMismatch {
+            method: "fetch_platform_state".to_string(),
+            known_versions: vec![0],
+            received: version,
+        })),
+    }
+}
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/storage/fetch_platform_state/mod.rs:29]: This should be a Platform method, it shouldn't be a global.

File: packages/rs-drive-abci/src/platform_types/platform_state/v0/mod.rs [hunk 7]
```diff
@@ -98,7 +99,7 @@ pub(super) struct PlatformStateForSavingV0 {
     /// Information about the last block
     pub last_committed_block_info: Option<ExtendedBlockInfo>,
     /// Current Version
-    pub current_protocol_version_in_consensus: ProtocolVersion,
+    pub current_protocol_version_in_consensus: ProtocolVersion, // TODO: Remove this
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/platform_types/platform_state/v0/mod.rs:102]: why would you want to remove this?

File: packages/rs-drive-abci/src/platform_types/platform_state/v0/mod.rs [hunk 8]
```diff
@@ -164,12 +165,17 @@ impl TryFrom<PlatformStateV0> for PlatformStateForSavingV0 {
...
+
+    fn try_from_platform_versioned(
+        value: PlatformStateForSavingV0,
+        platform_version: &PlatformVersion,
+    ) -> Result<Self, Self::Error> {
+        Ok(PlatformStateV0 {
             genesis_block_info: value.genesis_block_info,
             last_committed_block_info: value.last_committed_block_info,
-            current_protocol_version_in_consensus: value.current_protocol_version_in_consensus,
+            current_protocol_version_in_consensus: platform_version.protocol_version,
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/platform_types/platform_state/v0/mod.rs:178]: No, this seems wrong. You can't change the current protocol version in consensus.

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

File: packages/rs-drive-abci/src/execution/storage/fetch_platform_state/mod.rs (hunk 6 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/storage/fetch_platform_state/mod.rs:29]: There is no platform at the point we use it

File: packages/rs-drive-abci/src/platform_types/platform_state/v0/mod.rs (hunk 7 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/platform_types/platform_state/v0/mod.rs:102]: Oh, that's an artefact from our previous task when we were going to store platform state in trees

File: packages/rs-drive-abci/src/execution/storage/fetch_platform_state/mod.rs (hunk 6 above)
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/storage/fetch_platform_state/mod.rs:29]: It needs to be in the Platform code space.

Comment (shumkov) [reply packages/rs-drive-abci/src/execution/storage/fetch_platform_state/mod.rs:29]: It's already. Modules defines code space (namespaces). I don't think it's correct way to structure the code but if you insist I will add it to Platform struct.

File: packages/rs-drive-abci/src/execution/engine/initialization/init_chain/v0/mod.rs [hunk 9]
```diff
@@ -79,6 +79,13 @@ where
 
         state_guard.set_genesis_block_info(Some(genesis_block_info));
 
+        // Store initial protocol version
+        self.drive.store_current_protocol_version(
+            state_guard.current_protocol_version_in_consensus(),
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/engine/initialization/init_chain/v0/mod.rs:84]: won't this just be the default?

File: packages/rs-drive/src/drive/platform_state/fetch_platform_state_bytes/v0/mod.rs [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+use crate::error::Error;
+use grovedb::TransactionArg;
+
+impl Drive {
+    pub(super) fn fetch_platform_state_bytes_v0(
+        &self,
+        transaction: TransactionArg,
+    ) -> Result<Option<Vec<u8>>, Error> {
+        self.grove
+            .get_aux(EXECUTION_STORAGE_STATE_KEY, transaction)
```
Comment (QuantumExplorer) [inline packages/rs-drive/src/drive/platform_state/fetch_platform_state_bytes/v0/mod.rs:12]: rename EXECUTION_STORAGE_STATE_KEY

File: packages/rs-drive/src/drive/system/protocol_version/store_current_protocol_version.rs [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    ///
+    /// This function will return an error if the version of the Drive is unknown.
+    pub fn set_current_protocol_version_operations(
+        &self,
+        protocol_version: ProtocolVersion,
+        transaction: TransactionArg,
+        drive_operations: &mut Vec<LowLevelDriveOperation>,
+        drive_version: &DriveVersion,
+    ) -> Result<(), Error> {
+        self.batch_insert_if_changed_value(
```
Comment (QuantumExplorer) [inline packages/rs-drive/src/drive/system/protocol_version/store_current_protocol_version.rs:70]: As a general rule you shouldn't do things that require more ressources than required. This call for example would not be called very often but should be called when the protocol version changes. The if_changed_value is overkill here as we are sure based on business logic that the value has changed, and even if it hasn't the cost to make that change once per epoch is negligeable. While having if_changed_value doesn't break the system in any way it's a bad practice that is better to be discarded early.

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

File: packages/rs-drive-abci/src/execution/engine/initialization/init_chain/v0/mod.rs (hunk 9 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/engine/initialization/init_chain/v0/mod.rs:84]: Yeah, we should use genesis param instead of default.

File: packages/rs-drive/src/drive/platform_state/fetch_platform_state_bytes/v0/mod.rs [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+use crate::error::Error;
+use grovedb::TransactionArg;
+
+impl Drive {
+    pub(super) fn fetch_platform_state_bytes_v0(
+        &self,
+        transaction: TransactionArg,
+    ) -> Result<Option<Vec<u8>>, Error> {
+        self.grove
+            .get_aux(PLATFROM_STATE_KEY, transaction)
```
Comment (QuantumExplorer) [inline packages/rs-drive/src/drive/platform_state/fetch_platform_state_bytes/v0/mod.rs:12]: PLATFROM_STATE_KEY is mispelled

File: packages/rs-drive/src/drive/system/protocol_version/store_current_protocol_version.rs [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    ) -> Result<(), Error> {
+        self.batch_insert(
+            PathKeyElementInfo::PathFixedSizeKeyRefElement((
+                misc_path(),
+                PROTOCOL_VERSION_STORAGE_KEY,
+                Element::new_item(protocol_version.encode_var_vec()),
+            )),
+            drive_operations,
+            drive_version,
+        )?;
```
Comment (QuantumExplorer) [inline packages/rs-drive/src/drive/system/protocol_version/store_current_protocol_version.rs:76]: no need to have ? and then Ok

File: packages/rs-drive/src/drive/platform_state/fetch_platform_state_bytes/v0/mod.rs (hunk 12 above)
Comment (shumkov) [reply packages/rs-drive/src/drive/platform_state/fetch_platform_state_bytes/v0/mod.rs:12]: My favourite
